pep517/build.py

```diff
@@ -3,25 +3,50 @@
 import argparse
 import logging
 import os
-import contextlib
 import pytoml
 import shutil
-import errno
-import tempfile
 
 from .envbuild import BuildEnvironment
 from .wrappers import Pep517HookCaller
+from .dirtools import tempdir, mkdir_p
 
 log = logging.getLogger(__name__)
 
 
-@contextlib.contextmanager
-def tempdir():
-    td = tempfile.mkdtemp()
+def validate_system(system):
+    """
+    Ensure build system has the requisite fields.
+    """
+    required = {'requires', 'build-backend'}
+    if required > set(system):
+        missing = required - set(system)
+        message = "Missing required fields: {missing}".format(**locals())
+        raise ValueError(message)
+
+
+def load_system(source_dir):
+    """
+    Load the build system from a source dir (pyproject.toml).
+    """
+    pyproject = os.path.join(source_dir, 'pyproject.toml')
+    with open(pyproject) as f:
+        pyproject_data = pytoml.load(f)
+    return pyproject_data['build-system']
+
+
+def compat_system(source_dir):
+    """
+    Given a source dir, attempt to get a build system backend
+    and requirements from pyproject.toml. Fallback to
+    setuptools.
+    """
     try:
-        yield td
-    finally:
-        shutil.rmtree(td)
+        system = load_system(source_dir)
+    except Exception:
+        system = {}
+    system.setdefault('build-backend', 'setuptools.build_meta')
```

Comment:
I can see your point. My instinct was, since pep517 never had the legacy behavior, it's not necessary to drag it into here. That is, the "meta" command will never support that legacy assumption. That means it won't support those projects reliant on that behavior. If I recall, when `__legacy__` was created, it was meant for pip only (and that's why it's `__`-prefixed). I'd prefer to start with this constrained approach and relax the constraint later if demand and consideration indicate.

Reply:
I suspect we'll soon end up having to use the 'legacy' backend, because the people building a project through this mechanism are not necessarily the ones deciding what assumptions that project makes.

@pfmoore @gaborbernat (and anyone else): any preference on this?

Reply:
I'd go in +1 to make it setuptools.build_meta:__legacy__ the default.

Reply:
Updated in 205d128.